// trading-signals/dist/MAD/MAD.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterMAD = exports.MAD = void 0;
const __1 = require("..");
const Indicator_1 = require("../Indicator");
const util_1 = require("../util");
class MAD extends Indicator_1.BigIndicatorSeries {
    constructor(interval) {
        super();
        this.interval = interval;
        this.prices = [];
    }
    update(price) {
        this.prices.push(price);
        if (this.prices.length > this.interval) {
            this.prices.shift();
        }
        if (this.prices.length === this.interval) {
            return this.setResult(MAD.getResultFromBatch(this.prices));
        }
    }
    static getResultFromBatch(prices, average) {
        const mean = average || (0, util_1.getAverage)(prices);
        let sum = new __1.Big(0);
        for (let i = 0; i < prices.length; i++) {
            const deviation = new __1.Big(prices[i]).minus(mean).abs();
            sum = sum.plus(deviation);
        }
        return sum.div(prices.length || 1);
    }
}
exports.MAD = MAD;
class FasterMAD extends Indicator_1.NumberIndicatorSeries {
    constructor(interval) {
        super();
        this.interval = interval;
        this.prices = [];
    }
    update(price) {
        this.prices.push(price);
        if (this.prices.length > this.interval) {
            this.prices.shift();
        }
        if (this.prices.length === this.interval) {
            const mean = (0, util_1.getFasterAverage)(this.prices);
            let sum = 0;
            for (let i = 0; i < this.interval; i++) {
                const deviation = Math.abs(this.prices[i] - mean);
                sum += deviation;
            }
            return this.setResult(sum / this.interval);
        }
    }
    static getResultFromBatch(prices, average) {
        const mean = average || (0, util_1.getFasterAverage)(prices);
        let sum = 0;
        for (let i = 0; i < prices.length; i++) {
            const deviation = Math.abs(prices[i] - mean);
            sum += deviation;
        }
        return sum / prices.length;
    }
}
exports.FasterMAD = FasterMAD;
