// trading-signals/dist/DMA/DMA.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterDMA = exports.DMA = void 0;
const SMA_1 = require("../SMA/SMA");
class DMA {
    constructor(short, long, Indicator = SMA_1.SMA) {
        this.short = new Indicator(short);
        this.long = new Indicator(long);
    }
    get isStable() {
        return this.long.isStable;
    }
    update(price) {
        this.short.update(price);
        this.long.update(price);
    }
    getResult() {
        return {
            long: this.long.getResult(),
            short: this.short.getResult(),
        };
    }
}
exports.DMA = DMA;
class FasterDMA {
    constructor(short, long, SmoothingIndicator = SMA_1.FasterSMA) {
        this.short = new SmoothingIndicator(short);
        this.long = new SmoothingIndicator(long);
    }
    get isStable() {
        return this.long.isStable;
    }
    update(price) {
        this.short.update(price);
        this.long.update(price);
    }
    getResult() {
        return {
            long: this.long.getResult(),
            short: this.short.getResult(),
        };
    }
}
exports.FasterDMA = FasterDMA;
