// trading-signals/dist/MOM/MOM.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterMOM = exports.MOM = void 0;
const Indicator_1 = require("../Indicator");
const __1 = require("..");
const getFixedArray_1 = require("../util/getFixedArray");
class MOM extends Indicator_1.BigIndicatorSeries {
    constructor(interval) {
        super();
        this.interval = interval;
        this.historyLength = interval + 1;
        this.history = (0, getFixedArray_1.getFixedArray)(this.historyLength);
    }
    update(value) {
        this.history.push(value);
        if (this.history.length === this.historyLength) {
            return this.setResult(new __1.Big(value).minus(this.history[0]));
        }
    }
}
exports.MOM = MOM;
class FasterMOM extends Indicator_1.NumberIndicatorSeries {
    constructor(interval) {
        super();
        this.interval = interval;
        this.historyLength = interval + 1;
        this.history = (0, getFixedArray_1.getFixedArray)(this.historyLength);
    }
    update(value) {
        this.history.push(value);
        if (this.history.length === this.historyLength) {
            return this.setResult(value - this.history[0]);
        }
    }
}
exports.FasterMOM = FasterMOM;
