// trading-signals/dist/util/Period.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterPeriod = exports.Period = void 0;
const __1 = require("..");
const getFixedArray_1 = require("./getFixedArray");
const getMinimum_1 = require("./getMinimum");
const getMaximum_1 = require("./getMaximum");
class Period {
    constructor(interval) {
        this.interval = interval;
        this.values = (0, getFixedArray_1.getFixedArray)(interval);
    }
    getResult() {
        return {
            highest: this.highest,
            lowest: this.lowest,
        };
    }
    update(value) {
        this.values.push(new __1.Big(value));
        if (this.isStable) {
            this.lowest = (0, getMinimum_1.getMinimum)(this.values);
            this.highest = (0, getMaximum_1.getMaximum)(this.values);
            return this.getResult();
        }
    }
    get isStable() {
        return this.values.length === this.interval;
    }
}
exports.Period = Period;
class FasterPeriod {
    constructor(interval) {
        this.interval = interval;
        this.values = (0, getFixedArray_1.getFixedArray)(interval);
    }
    getResult() {
        return {
            highest: this.highest,
            lowest: this.lowest,
        };
    }
    update(value) {
        this.values.push(value);
        if (this.isStable) {
            this.lowest = Math.min(...this.values);
            this.highest = Math.max(...this.values);
            return this.getResult();
        }
    }
    get isStable() {
        return this.values.length === this.interval;
    }
}
exports.FasterPeriod = FasterPeriod;
