// trading-signals/dist/AO/AO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterAO = exports.AO = void 0;
const Indicator_1 = require("../Indicator");
const __1 = require("..");
const SMA_1 = require("../SMA/SMA");
class AO extends Indicator_1.BigIndicatorSeries {
    constructor(shortInterval, longInterval, SmoothingIndicator = SMA_1.SMA) {
        super();
        this.shortInterval = shortInterval;
        this.longInterval = longInterval;
        this.short = new SmoothingIndicator(shortInterval);
        this.long = new SmoothingIndicator(longInterval);
    }
    update({ low, high }) {
        const candleSum = new __1.Big(low).add(high);
        const medianPrice = candleSum.div(2);
        this.short.update(medianPrice);
        this.long.update(medianPrice);
        if (this.long.isStable) {
            return this.setResult(this.short.getResult().sub(this.long.getResult()));
        }
    }
}
exports.AO = AO;
class FasterAO extends Indicator_1.NumberIndicatorSeries {
    constructor(shortInterval, longInterval, SmoothingIndicator = SMA_1.FasterSMA) {
        super();
        this.shortInterval = shortInterval;
        this.longInterval = longInterval;
        this.short = new SmoothingIndicator(shortInterval);
        this.long = new SmoothingIndicator(longInterval);
    }
    update({ low, high }) {
        const medianPrice = (low + high) / 2;
        this.short.update(medianPrice);
        this.long.update(medianPrice);
        if (this.long.isStable) {
            return this.setResult(this.short.getResult() - this.long.getResult());
        }
    }
}
exports.FasterAO = FasterAO;
