// trading-signals/dist/Indicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumberIndicatorSeries = exports.BigIndicatorSeries = void 0;
const error_1 = require("./error");
class BigIndicatorSeries {
    get isStable() {
        return this.result !== undefined;
    }
    getResult() {
        if (this.result === undefined) {
            throw new error_1.NotEnoughDataError();
        }
        return this.result;
    }
    setResult(value) {
        if (this.highest === undefined || value.gt(this.highest)) {
            this.highest = value;
        }
        if (this.lowest === undefined || value.lt(this.lowest)) {
            this.lowest = value;
        }
        return (this.result = value);
    }
}
exports.BigIndicatorSeries = BigIndicatorSeries;
class NumberIndicatorSeries {
    get isStable() {
        return this.result !== undefined;
    }
    getResult() {
        if (this.result === undefined) {
            throw new error_1.NotEnoughDataError();
        }
        return this.result;
    }
    setResult(value) {
        if (this.highest === undefined || value > this.highest) {
            this.highest = value;
        }
        if (this.lowest === undefined || value < this.lowest) {
            this.lowest = value;
        }
        return (this.result = value);
    }
}
exports.NumberIndicatorSeries = NumberIndicatorSeries;
