// trading-signals/dist/ATR/ATR.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterATR = exports.ATR = void 0;
const Indicator_1 = require("../Indicator");
const TR_1 = require("../TR/TR");
const WSMA_1 = require("../WSMA/WSMA");
class ATR extends Indicator_1.BigIndicatorSeries {
    constructor(interval, SmoothingIndicator = WSMA_1.WSMA) {
        super();
        this.interval = interval;
        this.tr = new TR_1.TR();
        this.smoothing = new SmoothingIndicator(interval);
    }
    update(candle) {
        const trueRange = this.tr.update(candle);
        this.smoothing.update(trueRange);
        if (this.smoothing.isStable) {
            return this.setResult(this.smoothing.getResult());
        }
    }
}
exports.ATR = ATR;
class FasterATR extends Indicator_1.NumberIndicatorSeries {
    constructor(interval, SmoothingIndicator = WSMA_1.FasterWSMA) {
        super();
        this.interval = interval;
        this.tr = new TR_1.FasterTR();
        this.smoothing = new SmoothingIndicator(interval);
    }
    update(candle) {
        const trueRange = this.tr.update(candle);
        this.smoothing.update(trueRange);
        if (this.smoothing.isStable) {
            return this.setResult(this.smoothing.getResult());
        }
    }
}
exports.FasterATR = FasterATR;
