// trading-signals/dist/MACD/MACD.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterMACD = exports.MACD = void 0;
const __1 = require("..");
class MACD {
    constructor(config) {
        this.prices = [];
        this.long = new config.indicator(config.longInterval);
        this.short = new config.indicator(config.shortInterval);
        this.signal = new config.indicator(config.signalInterval);
    }
    get isStable() {
        return this.result !== undefined;
    }
    update(_price) {
        const price = new __1.Big(_price);
        this.prices.push(price);
        const short = this.short.update(price);
        const long = this.long.update(price);
        if (this.prices.length > this.long.interval) {
            this.prices.shift();
        }
        if (this.prices.length === this.long.interval) {
            const macd = short.sub(long);
            const signal = this.signal.update(macd);
            return (this.result = {
                histogram: macd.sub(signal),
                macd: macd,
                signal,
            });
        }
    }
    getResult() {
        if (!this.isStable || this.result === undefined) {
            throw new __1.NotEnoughDataError();
        }
        return this.result;
    }
}
exports.MACD = MACD;
class FasterMACD {
    constructor(short, long, signal) {
        this.short = short;
        this.long = long;
        this.signal = signal;
        this.prices = [];
    }
    getResult() {
        if (this.result === undefined) {
            throw new __1.NotEnoughDataError();
        }
        return this.result;
    }
    get isStable() {
        return this.result !== undefined;
    }
    update(price) {
        this.prices.push(price);
        const short = this.short.update(price);
        const long = this.long.update(price);
        if (this.prices.length > this.long.interval) {
            this.prices.shift();
        }
        if (this.prices.length === this.long.interval) {
            const macd = short - long;
            const signal = this.signal.update(macd);
            return (this.result = {
                histogram: macd - signal,
                macd,
                signal,
            });
        }
    }
}
exports.FasterMACD = FasterMACD;
