// trading-signals/dist/DEMA/DEMA.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterDEMA = exports.DEMA = void 0;
const EMA_1 = require("../EMA/EMA");
const Indicator_1 = require("../Indicator");
class DEMA extends Indicator_1.BigIndicatorSeries {
    constructor(interval) {
        super();
        this.interval = interval;
        this.inner = new EMA_1.EMA(interval);
        this.outer = new EMA_1.EMA(interval);
    }
    update(price) {
        const innerResult = this.inner.update(price);
        const outerResult = this.outer.update(innerResult);
        return this.setResult(innerResult.times(2).sub(outerResult));
    }
    get isStable() {
        return this.outer.isStable;
    }
}
exports.DEMA = DEMA;
class FasterDEMA extends Indicator_1.NumberIndicatorSeries {
    constructor(interval) {
        super();
        this.interval = interval;
        this.inner = new EMA_1.FasterEMA(interval);
        this.outer = new EMA_1.FasterEMA(interval);
    }
    update(price) {
        const innerResult = this.inner.update(price);
        const outerResult = this.outer.update(innerResult);
        return this.setResult(innerResult * 2 - outerResult);
    }
    get isStable() {
        return this.outer.isStable;
    }
}
exports.FasterDEMA = FasterDEMA;
