// trading-signals/dist/CG/CG.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FasterCG = exports.CG = void 0;
const Indicator_1 = require("../Indicator");
const __1 = require("..");
const SMA_1 = require("../SMA/SMA");
class CG extends Indicator_1.BigIndicatorSeries {
    get isStable() {
        return this.signal.isStable;
    }
    constructor(interval, signalInterval) {
        super();
        this.interval = interval;
        this.signalInterval = signalInterval;
        this.prices = [];
        this.signal = new SMA_1.SMA(signalInterval);
    }
    update(price) {
        this.prices.push(new __1.Big(price));
        if (this.prices.length > this.interval) {
            this.prices.shift();
        }
        let nominator = new __1.Big(0);
        let denominator = new __1.Big(0);
        for (let i = 0; i < this.prices.length; ++i) {
            const price = this.prices[i];
            nominator = nominator.plus(price.mul(i + 1));
            denominator = denominator.plus(price);
        }
        const cg = denominator.gt(0) ? nominator.div(denominator) : new __1.Big(0);
        this.signal.update(cg);
        if (this.signal.isStable) {
            return this.setResult(cg);
        }
    }
}
exports.CG = CG;
class FasterCG extends Indicator_1.NumberIndicatorSeries {
    get isStable() {
        return this.signal.isStable;
    }
    constructor(interval, signalInterval) {
        super();
        this.interval = interval;
        this.signalInterval = signalInterval;
        this.prices = [];
        this.signal = new SMA_1.FasterSMA(signalInterval);
    }
    update(price) {
        this.prices.push(price);
        if (this.prices.length > this.interval) {
            this.prices.shift();
        }
        let nominator = 0;
        let denominator = 0;
        for (let i = 0; i < this.prices.length; ++i) {
            const price = this.prices[i];
            nominator = nominator + price * (i + 1);
            denominator = denominator + price;
        }
        const cg = denominator > 0 ? nominator / denominator : 0;
        this.signal.update(cg);
        if (this.signal.isStable) {
            return this.setResult(cg);
        }
    }
}
exports.FasterCG = FasterCG;
